src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+                             ClusterId clusterId)
+{
+    BindingTableEntry * entry =
+        mBindingEntries.CreateObject(sourceEndpoint, peerNodeId, peerEndpointId, peerFabricIndex, clusterId);
+    if (entry == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    if (FindPeerDevice(entry->mPeerId) == nullptr)
+    {
+        DeviceProxyInitParams initParams = { &Server::GetInstance().GetSecureSessionManager(),
+                                             &Server::GetInstance().GetExchangeManager(),
+                                             &Server::GetInstance().GetSessionIDAllocator(),
+                                             GetFabricInfoFromIndex(peerFabricIndex) };
+        OperationalDeviceProxy * device  = mPeerDevices.CreateObject(initParams, entry->mPeerId);
+        if (device == nullptr)
+        {
+            mBindingEntries.ReleaseObject(entry);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return TryConnectDevice(device);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::Remove(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId,
+                                FabricIndex peerFabricIndex, ClusterId clusterId)
+{
+    BindingTableEntry * removeEntry = nullptr;
+    int peerNumBindings             = 0;
+    PeerId peer                     = BuildPeerId(peerNodeId, peerFabricIndex);
+    mBindingEntries.ForEachActiveObject(
+        [&removeEntry, &peerNumBindings, sourceEndpoint, peer, peerEndpointId, clusterId](BindingTableEntry * entry) {
+            if (entry->mSourceEndpoint == sourceEndpoint && entry->mPeerId == peer && entry->mPeerEndpointId == peerEndpointId &&
+                entry->mClusterId == clusterId)
+            {
+                removeEntry = entry;
+            }
+            if (entry->mPeerId == peer)
+            {
+                peerNumBindings++;
+            }
+            return true;
+        });
+    if (removeEntry == nullptr)
+    {
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+    mBindingEntries.ReleaseObject(removeEntry);
+    // Last binding with the peer, close the connection.
+    if (peerNumBindings == 1)
+    {
+        OperationalDeviceProxy * device = FindPeerDevice(peer);
+        if (device)
+        {
+            mPeerDevices.ReleaseObject(device);
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::TryConnectDevice(OperationalDeviceProxy * device)
+{
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
```

Comment:
So if this fails, what happens with `device`?  Who frees it?

Reply:
The `device` won't be freed on failures. It's freed by user by unbinding the device.

Reply:
OK, but it's not connected either.  It just sits there, completely unusable, wasting memory.  Why?

If the intent is to do something in terms of trying to connect it later, etc, that should be documented with TODO comments.